### ISSUE TITLE
Interop flake: cloud_to_cloud:go:java_server:timeout_on_sleeping_server

### ISSUE THREAD

**Issue by apolcyn** [grpc/grpc#15745]
Noticed this on PR: https://github.com/grpc/grpc/pull/15490

```
2018-06-13 11:21:44,047 FATAL: 2018/06/13 18:21:43 &{0xc42026a080}.Recv() = _, rpc error: code = Canceled desc = stream terminated by RST_STREAM with error code: CANCEL, want error code 4
exit status 1

2018-06-13 11:21:44,048 FAILED: cloud_to_cloud:go:java_server:timeout_on_sleeping_server [ret=1, pid=5975, time=2.1sec]
```

https://source.cloud.google.com/results/invocations/b24dc172-cb1d-4293-a364-0f4db5dd4fa8/targets

**Comment by ejona86** [comment]
It seems the server noticed the deadline before the client. @dfawley, does the client double-check the deadline when the response is CANCELLED?

**Comment by dfawley** [comment]
tl;dr: There is a race here, we do have a place where we could catch it, but it is not 100% reliable.  Also, note that this race is impossible to fully avoid in the real world due to clock skew (i.e. frequency, not absolute time).

Details:

When the RST_STREAM comes, the stream is immediately closed and CANCELLED is latched into its status:

https://github.com/grpc/grpc-go/blob/107f19369e55f2361cb2f6bfcf27bd3385eec74a/transport/http2_client.go#L909

This races with a goroutine blocked on the stream's context, which will wake up when the deadline is reached:

https://github.com/grpc/grpc-go/blob/107f19369e55f2361cb2f6bfcf27bd3385eec74a/stream.go#L327-L328

Reading from the stream _could_ still return the deadline exceeded here, but it's in a select, so it's only 50% likely to hit:

https://github.com/grpc/grpc-go/blob/24f3cca1ffddeab04708c325f8088c76c2d74972/transport/transport.go#L143

I think it would be fine to do a double-check here:
https://github.com/grpc/grpc-go/blob/24f3cca1ffddeab04708c325f8088c76c2d74972/transport/transport.go#L148

So that any read that gets a CANCELLED error would get a DEADLINE_EXCEEDED result instead.

**Comment by ejona86** [comment]
> Also, note that this race is impossible to fully avoid in the real world due to clock skew (i.e. frequency, not absolute time).

I agree, but the skew has to be bad enough to overcome 1 RTT + 2x server processing + 2x client processing. That's more likely if you have a really big deadline, but still seems pretty rare.

For reference, this is where Java does the double-check:
https://github.com/grpc/grpc-java/blob/v1.12.0/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L569

C does something similar. This seems a "Go didn't get the memo" bug. The race happened quite a bit in Java and C without the double-checking logic.

It looks like you should call context.Context.Deadline() when you double-check, as that would not be influenced by goroutine scheduling.
